boa/src/syntax/lexer/string.rs

```diff
@@ -58,171 +58,270 @@ impl<R> Tokenizer<R> for StringLiteral {
-                                _ => (),
+                        }
+                        _ if escape_ch.is_digit(8) => {
+                            Self::take_legacy_octal_escape_sequence(
+                                cursor,
+                                ch_start_pos,
+                                Some(&mut buf),
+                                strict_mode,
+                                escape_ch as u8,
+                            )?;
+                        }
+                        _ if Self::is_line_terminator(escape_ch) => {
+                            // Grammar: LineContinuation
+                            // Grammar: \ LineTerminatorSequence
+                            // LineContinuation is the empty String. Do nothing and continue lexing.
+                        }
+                        _ => {
+                            if escape_ch.len_utf16() == 1 {
+                                buf.push(escape_ch as u16);
+                            } else {
+                                buf.extend(escape_ch.encode_utf16(&mut [0u16; 2]).iter());
                             }
-                            buf.push(o as u16);
                         }
-                        _ => buf.push(escape as u16),
                     };
                 }
+                Some(ch) => {
+                    if ch.len_utf16() == 1 {
+                        buf.push(ch as u16);
+                    } else {
+                        buf.extend(ch.encode_utf16(&mut [0u16; 2]).iter());
+                    }
+                }
+                None => {
+                    return Err(Error::from(io::Error::new(
+                        ErrorKind::UnexpectedEof,
+                        "unterminated string literal",
+                    )));
+                }
             }
-            Some(next_ch) => {
-                if next_ch.len_utf16() == 1 {
-                    buf.push(next_ch as u16);
-                } else {
-                    let mut code_point_bytes_buf = [0u16; 2];
-                    let code_point_bytes = next_ch.encode_utf16(&mut code_point_bytes_buf);
+        }
 
-                    buf.extend(code_point_bytes.iter());
+        Ok((
+            String::from_utf16_lossy(buf.as_slice()),
+            Span::new(start_pos, cursor.pos()),
+        ))
+    }
+
+    #[inline]
+    pub(super) fn take_unicode_escape_sequence<R>(
+        cursor: &mut Cursor<R>,
+        start_pos: Position,
+        code_units_buf: Option<&mut Vec<u16>>,
+    ) -> Result<u32, Error>
+    where
+        R: Read,
+    {
+        // Support \u{X..X} (Unicode CodePoint)
+        if cursor.next_is(b'{')? {
+            // TODO: use bytes for a bit better performance (using stack)
+            let mut code_point_buf = Vec::with_capacity(6);
+            cursor.take_until(b'}', &mut code_point_buf)?;
+
+            let code_point_str = unsafe { str::from_utf8_unchecked(code_point_buf.as_slice()) };
```

Comment:
I added a comment for it. Yes it is a little bit unclear why it's safe to do this.

Reply:
It is probably safe because [`u32::from_str_radix`][impl] does not make the assumption that the string is valid UTF-8. It treats it as ASCII, and if any character is not an ASCII digit, then it safely returns an error. As `code_point_str` is used only for calling this function, this is probably safe.
Nevertheless, I think another maintainer should double-check.

[impl]: https://doc.rust-lang.org/src/core/num/mod.rs.html#815-839

Reply:
Here is a sample: 
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8871829a649119c020e55943ee92df76

Reply:
Yes, it confirms what I thought. What do you think @RageKnify? If it is OK for you, then it is also fine for me.

Reply:
I'm not sure that this is safe, functions that receive `&str` assume `str` to be valid utf8, `u32::from_str_radix` makes no promises in regards to what it might do if it's argument is not valid utf8, I don't think what we're doing is safe.

Reply:
@RageKnify Yes, there is no promise from the doc. For the sake of safety, I changed the conversion to safe byte-to-string.